Include/cpython/pystate.h

```diff
@@ -200,6 +200,9 @@ struct _ts {
        The PyThreadObject must hold the only reference to this value.
     */
     PyObject *threading_local_sentinel;
+
+    /* Set when the thread is about to exit */
+    struct _PyEventRc *thread_is_exiting;
```

Comment:
Let's see if I've understood the situation.

`ThreadHandle`:

* is strictly internal to the _thread module
* its lifetime is tied to its refcount
* the following have a reference:
    * each `PyThreadHandleObject` (bound to each `threading.Thread`)
    * `thread_PyThread_start_new_thread()`
    * `thread_run()`
    * the "boot state"
    * `thread_shutdown()` (temporary)
* the following hold a pointer to a linked-list node for each handle but not a reference to the handle:
    * `_PyThread.threads.handles` (`ThreadHandle.node`) - from creation to destruction
    * the module state's `shutdown_handles` (**only if not daemon**) - while running

The "exiting" event:

* currently stored in the non-pointer `ThreadHandle.thread_is_exiting` field
* not directly tied to `ThreadHandle` lifetime
* notified by:
    * `thread_run()` at the end
    * `_PyThread_AfterFork()` for each handle in `_PyRuntime.threads.handles`
    * `_thread._ThreadHandle._set_done()`
    * `ThreadHandle_start()` when failed to start

The change in this PR is:

* start notifying the event in `PyThreadState_Clear()`
* this requires we keep a reference to the event on `PyThreadState`
* the dualing lifetimes for `ThreadHandle` and `PyThreadState` require that the event exist independently of the handle
* this requires a refcount for the event

Reply:
Some very mildly related questions that don't demand an answer here 😄:

* could we generalize `ThreadHandle` and move it to (incorporate into) Python/thread.c?
* why do we have `_PyRuntime.threads.handles` instead of `PyInterpreterState.threads.handles`?

Reply:
> Let's see if I've understood the situation.

Yep, that's accurate!

> could we generalize ThreadHandle and move it to (incorporate into) Python/thread.c

Potentially, but that's a pretty big change when we're close to the release date for 3.13. I'd prefer to aim for a targeted fix and we can tackle a larger refactoring later if desired. I think the best solution here is probably what you suggest in a later comment: store a reference to `ThreadHandle` on `PyThreadState`. I was initially hesitant to do that since `ThreadHandle` is intended to be internal, but I think I can stick the reference in `_PyThreadStateImpl` and achieve the same effect. I'll put up a separate PR for that approach.

> why do we have _PyRuntime.threads.handles instead of PyInterpreterState.threads.handles?

Because they're used post-fork, which affects the entire process.